lib/fuzzer/InputPoolPosix.zig

```diff
@@ -0,0 +1,168 @@
+// file mapped by all fuzzing processes that stores the end index of every
+// stored string and some metadata (see the meta field).
+//
+// When a process finds a new good input, it (under a lock) appends to both of
+// these files. All other processes can immediately start working on this input
+//
+// We never shrink the corpus when fuzzing. One has to stop fuzzing and run a
+// separate program. That is because erasing strings would require complicated
+// synchronization between fuzzing processes. Finding good inputs seems to be
+// rare enough for the input pool to never get too large.
+
+const std = @import("std");
+const assert = std.debug.assert;
+const util = @import("util.zig");
+const check = util.check;
+const MemoryMappedList = @import("memory_mapped_list.zig").MemoryMappedList;
+
+/// maximum 2GiB of input data should be enough. 32th bit is delete flag
+pub const Index = u31;
+
+const InputPoolPosix = @This();
+
+const LatestFormatVersion: u8 = 0;
+const SignatureVersion: u32 = @bitCast([4]u8{ 'V', 'N', 'S', LatestFormatVersion });
+
+/// mmap-ed file. In reality we mapped 2 GiB but cropped to match the mmaped
+/// file size. Doesn't need mremap when growing.
+buffer: MemoryMappedList(u8),
+
+/// mmap-ed file. In reality we mapped 2 GiB but cropped to match the mmaped
+/// file size. Doesn't need mremap when growing.
+///
+/// layout of the meta file (v0):
+/// [0] [3]u8 file signature + u8 format version
+/// [1] u32 for xcmpchg (mutex in the stdlib is not FUTEX_SHARED)
+/// [2] u32 deleted bytes
+/// [3] u32 number of strings
+/// [4..] data (string end offsets)
+meta: MemoryMappedList(u32),
+
+pub fn init(dir: std.fs.Dir, pc_digest: u64) InputPoolPosix {
+    const hex_digest = std.fmt.hex(pc_digest);
+
+    const buffer_file_path = "v/" ++ hex_digest ++ "buffer";
+    const meta_file_path = "v/" ++ hex_digest ++ "meta";
+    const buffer_file = check(@src(), dir.createFile(buffer_file_path, .{
+        .read = true,
+        .truncate = false,
+    }), .{ .file = buffer_file_path });
+    const meta_file = check(@src(), dir.createFile(meta_file_path, .{
+        .read = true,
+        .truncate = false,
+    }), .{ .file = meta_file_path });
+
+    const buffer = MemoryMappedList(u8).init(buffer_file, std.math.maxInt(Index));
+    var meta = MemoryMappedList(u32).init(meta_file, std.math.maxInt(Index));
+
+    if (meta.items.len == 0) {
+        meta.appendSlice(&.{
+            SignatureVersion, // signature
+            Unlocked, // mutex
+            0, // deleted bytes
+            0, // number of strings
+        });
+    } else {
+        assert(meta.items[0] == SignatureVersion);
+        assert(meta.items[1] == Unlocked or meta.items[1] == Locked);
+    }
+
+    return .{
+        .buffer = buffer,
+        .meta = meta,
+    };
+}
+
+pub fn deinit(ip: InputPoolPosix) void {
+    ip.buffer.deinit();
+    ip.meta.deinit();
+}
+
+// Primitive spin lock implementation. There is basically no contention on it.
```

Comment:
Use `std.Thread.Mutex` instead of implementing a spin lock.

Reply:
The problem with that was that `std.Thread.Mutex` calls `futex_wait` with `PRIVATE_FLAG`, which signals to the kernel that this futex is not shared with other processes, which is not the case probably. Depends on how we are going to implement paralell fuzzing. I wanted to ask you for input on that.

Reply:
I was thinking that the build system might spawn multiple separate fuzz processes and in that case, this lock would be inter-process and the `PRIVATE_FLAG` would be problematic. It would not be a problem if the fuzzer itself just spawns a couple of threads in a thread pool.

I dont know that is the policy of spawning threads manually vs letting the build system handle parallelism.

Reply:
I think it will be better to let the build system handle parallelism because:
* I didn't get #20274 working in my most recent effort and I lack the motivation to work on that problem currently because it involves sad workarounds for a disappointing lack of primitives offered by Unix-like operating systems. Since the thread pool is in the build runner process, the fuzzer process does not know how many threads it can spawn.
* Eventually, we want the ability to coordinate multiple build runners fuzzing across multiple different machines. This means coordinating seeds and inputs at that higher abstraction layer.

If we need to add an inter-process mutex to the standard library, that's fine, let's do it.

However there is also the lock-free approach. I haven't closely inspected what the mutex in this PR is being used for yet. Maybe locking could be entirely avoided.

Reply:
> However there is also the lock-free approach.

I failed to find one that works. The main thing that is done under the lock is a `ftruncate` syscall and I don't see how to avoid that.

> If we need to add an inter-process mutex to the standard library, that's fine, let's do it.

I don't like the idea of adding a new mutex entierly since the only difference from the current mutex is that it passes different flags to a syscall. Imo there should be a compile time configuration of the `Mutex` struct but that is a nontrivial change probably worth entire issue/PR on its own. I will declare the fuzzer single-process for now and address this in the next iteration.

My main reason for this is to keep this PR small.